network_simulator/flow.py

```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+                                                      FLOW_TIMEOUT_SEC)
+
+
+
+class FlowTimeoutPacketEvent(Event):
+    def __init__(self, flow, packet):
+        """
+        Triggered when packet has timeout at a flow.
+
+        :ivar Flow flow: flow of this event.
+        :ivar packet packet: the particular packet of this timeout.
+        """
+        self.flow = flow
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        # Does nothing.
+        pass
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule DeviceToLinkEvent and renew FlowTimeoutPacketEvent for each
+        packet.
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # If the particular packet is currently in transit, then we don't
+        # need to do anything here.
+        if self.packet.packet_id not in self.flow.packets_in_transit:
+            return
+        flow_send_event = FlowSendPacketsEvent(self.flow, [self.packets])
+        main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
+
+
+
+class FlowReceivedAckEvent(Event):
+    """
+    Triggered when an ACK is received by a flow.
+    """
+    def __init__(self, flow, packet):
+        """
+        :ivar Flow flow: flow of this event.
+        :ivar AckPacket packet: ack packet received.
+        """
+        self.flow = flow
+        self.packet = packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Triggered whenever an ACK is received.
+
+        :param MainEventLoop main_event_loop: main loop to retrieve the
+        simulation time of function call.
+        :param Statistics statistics: the Statistics to update.
+        """
+        # Update RTT.
+        sent_time = self.packet.data_packet_start_time_sec
+        curr_rtt = main_event_loop.global_clock_sec - sent_time
+        self.flow.packet_rtts.append((self.packet.packet_id, curr_rtt))
+
+        # Remove packet in transit.
+        self.flow.packets_in_transit.remove(self.packet.packet_id)
+
+        # TODO(sharon): Base on following description, update the code when
+        # we have finished TCP algorithm setup.
+        # 1. If there is a gap in the list of sent packets by ACK, then that
+        # means that a packet was lost (since we are using selective repeat).
+        # Update window_size accordingly based on TCP algorithm.
+        # 2. If there is a duplicate ACK, update window_size and other params
+        # based on TCP algorithm.
+        # 3. If there is no gap, then there was no packet loss. Update
+        # window_size based on TCP algorithm.
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule Flow event based on packet loss.
+
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # gap = a packet loss (gap in ACK or dupACK)
```

Comment:
flow.handle_packet_loss() should automatically deal with packet losses and regenerating FlowSendPacketEvents. This is probably cleaner since conditions for retransmit can get complicated.

So you should only do something in schedule_new_events() if had_packet_loss is False.